stdlib/public/SwiftShims/RefCount.h

```diff
@@ -725,7 +747,27 @@ class RefCounts {
     } while (!refCounts.compare_exchange_weak(oldbits, newbits,
                                               std::memory_order_relaxed));
   }
-
+  
+  void setHasNoObjCComplications(bool nonobjc) {
+    auto oldbits = refCounts.load(SWIFT_MEMORY_ORDER_CONSUME);
+    //Immortal and no objc complications share a bit, so don't let setting
+    //the complications one clear the immmortal one
+    if (oldbits.isImmortal() || oldbits.hasNoObjCComplications() == nonobjc) {
+      return;
+    }
+    RefCountBits newbits;
+    do {
+      newbits = oldbits;
+      newbits.setHasNoObjCComplications(nonobjc);
+    } while (!refCounts.compare_exchange_weak(oldbits, newbits,
+                                              std::memory_order_relaxed));
```

Comment:
Don't you also need to clear the field when you spill into the slow representation, to avoid being treated as immortal?

Reply:
It's a bit convoluted to read, but unless I'm missing something the code that creates the side table doesn't use any bits from the old refcount for its new value (it does use them for the pointed-to value)

```
  HeapObjectSideTableEntry *side = new HeapObjectSideTableEntry(getHeapObject());
  auto newbits = InlineRefCountBits(side);
  ...
    } while (! refCounts.compare_exchange_weak(oldbits, newbits,
                                             std::memory_order_release,
                                             std::memory_order_relaxed));
```

and the constructor it's using there is

```
  LLVM_ATTRIBUTE_ALWAYS_INLINE
  RefCountBitsT(HeapObjectSideTableEntry* side)
    : bits((reinterpret_cast<BitsType>(side) >> Offsets::SideTableUnusedLowBits)
           | (BitsType(1) << Offsets::UseSlowRCShift)
           | (BitsType(1) << Offsets::SideTableMarkShift))
  {
    assert(refcountIsInline);
  }
```

Reply:
Ah, okay. That makes sense.

Reply:
Does this mean `incrementStrongExtraRefCount` doesn't need adjustment for its `This deliberately overflows into the UseSlowRC field.` with the new scheme?

Reply:
It shouldn't need any changes, the increment should overflow into that field, and then the swap replaces it. I want to look very very carefully at this today though, because this is the most likely place for it to be failing those tests (I see no failures locally).

hm…

I guess if it overflows into UseSlowRC there will be a brief period where the object thinks it's immortal and contending refcounting gets dropped? How does the current scheme handle that?

Reply:
Oh it handles it by pushing it onto the slow path where it tries to allocate a side table and loses the race. Right. Ok, this is probably the bug then.